### ISSUE TITLE
Bunch of issues

### ISSUE THREAD

**Issue by JustM57** [catboost/catboost#229]
I'll switch to russian.
Должен сказать, что мы активно пользуемся вашей библиотекой. Вы меня в этом убедили после презентации(туториала) а офисе Яндекса. Но есть куча странных багов, которые непонятно откуда берутся.
Попробовали мы посчитать модель на гпушке. Конечно немного обидно, что не показываются графики, но с надеждой на ускорение обучения мы попробовали этот вариант. Чтобы не заспамить чат кучей сообщений об обученных деревьях verbose = Silent. Соответственно узнать как долго и как хорошо перформит модель нельзя. Для подобной визуализации может подойти tqdm если бы была возможность дообучать модель. Проблема не в этом, проблема в том, что когда модель посчитана и её нужно сохранить, то она съела 15 гб оперативки. Я сначала не понял что к чему, поэтому загрузил второй юпитер ноутбук, в котором есть возможность визуализации любой кастом метрики через model.shrink(код могу сбросить). Во время загрузки модели все 31гб оперативки компьютера закончились.. В то время как ЦПУ модель мало того, что показывает графики, хотя и учится раз в 3 медленнее, но весит 10мб. Одна и та же модель по идее... Ещё очень странные предсказания у гпу модели(валидация), в то время как цпу модель выдаёт белый шум вокруг 0.5, гпу выдаёт нечто смещённое
![gpu_model](https://user-images.githubusercontent.com/16518742/35219191-b267a46c-ff82-11e7-8ab6-9bdf70b29527.png)
Теперь про кастомную метрику. У нас получилось реализовать её через Pool c заданными весами, но есть несколько проблем:
Она была выбрана вы качестве eval metrics, параметры od_wait и od_type прописаны как всегда, но модель перестаёт учится в случайный момент. Там явно нет никаких 500 итераций после последнего минимума.
Loss function, который написан у вас в примере не работает. Просто выдаёт нечто константное.
Не совсем понятно как соотвествует модель score и аккураси. Прикладываю картинку:
![screenshot from 2018-01-22 14-07-55](https://user-images.githubusercontent.com/16518742/35219411-90e79e40-ff83-11e7-95cb-d373ee3f4c68.png)

**Comment by Noxoomo** [comment]
Привет!

1) Можешь написать версии катбуста и параметры, с которыми запускались GPU/CPU.
Плюс статистики пула (сколько признаков, сколько вещественных/категориальных, число документов, есть ли веса в датасете). 
И размеры моделей GPU/CPU (количество деревьев)
2)  logging_level=Info будет писать только ошибки без информации о деревьях. 
3)GPU не умеет считать кастомные метрики, а вот ошибку для оптимизируемой loss-function пишет (и должен визуализировать, если не показывает, то поправим)
4)  Вместо model.shrink  можно делать stage_predict (а для существующих метрик вызвать eval_metrics)

**Comment by JustM57** [comment]
@Noxoomo 
1. 0.5.2.1

```kitty = cat.CatBoostClassifier(
    iterations = 20000,
    learning_rate = 0.001,
    depth = 8,
    l2_leaf_reg = 8,
    #bootstrap_type = 'Bayesian',
    #bagging_temperature = 8,
    random_strength = 8,
    #eval_metric = PNLMetric(returns = X_test['Si.returns.close'][1:], n=15, proba=False, commission = True),
    custom_loss = ['Accuracy', 'AUC'],
                  #PNLMetric(returns = X_test['Si.returns.close'][1:], n=15, proba=False, commission = True)],
    random_seed = 57,
    logging_level = 'Silent',
    use_best_model = False,
    calc_feature_importance=True,
    od_type = 'Iter',
    od_wait = 2000,
    counter_calc_method = 'SkipTest',
    has_time = True
)

kitty.fit(
    X_train, y_train,
    cat_features = categorical_features_indices,
    eval_set=(X_test, y_test),
    #logging_level='Verbose',  # you can uncomment this for text output
    plot=False
)```

Около 300 фичей, практически все числовые. В датасете не было весов

**Comment by JustM57** [comment]
Still no idea how does test accuracy plot and actual model score is correlated. I used data in pool with weights, but metrics was deafult for classification(logloss)
![2018-01-23 14-31-42](https://user-images.githubusercontent.com/16518742/35273692-8b5cba42-004a-11e8-8bfe-9362a546f5b0.png)

**Comment by JustM57** [comment]
true plot:
![2018-01-23 15-01-54](https://user-images.githubusercontent.com/16518742/35274758-66a5eff8-004e-11e8-8a05-e2395c4ac906.png)

**Comment by JustM57** [comment]
akward

**Comment by Noxoomo** [comment]
> Проблема не в этом, проблема в том, что когда модель посчитана и её нужно сохранить, то она съела 15 гб оперативки. Я сначала не понял что к чему, поэтому загрузил второй юпитер ноутбук, в котором есть возможность визуализации любой кастом метрики через model.shrink(код могу сбросить). Во время загрузки модели все 31гб оперативки компьютера закончились.. В то время как ЦПУ модель мало того, что показывает графики, хотя и учится раз в 3 медленнее, но весит 10мб.


Про размер модели: если обучалось 20000 деревьев, то модель на 10MB выглядит как что-то обрезанное. 20к деревьев даже без категориальных факторов будет занимать больше.

Без данных и логов диагностировать проблему сложно: можешь обучить GPU и CPU (с самой последней версией пакета) с уровнем логирования Info и показать лог обучения?

**Comment by JustM57** [comment]
@Noxoomo на моём компьютере нет такой возможности. Как только снова будет возможность провести эксперимент - сделаю.
PS Там не 10мб, да, 60мб - но порядок такой же... Явно не 15 гигов оперативки

**Comment by Noxoomo** [comment]
А модель на диске тоже 15GB заняла? И сколько строчек в датасете.
И еще: на GPU и на CPU оба раза были eval_metrics пустыми? У тебя есть опции 
```
od_type = 'Iter',
 od_wait = 2000,
```
которые останавливают обучение по eval_metric'е
В случае GPU это всегда logloss для классификации, а на CPU то, что указано в опции.
Если здесь были отличия, то CPU мог обучить не 20к деревьев, а меньше.

**Comment by JustM57** [comment]
@Noxoomo сколько на диске занимает гпу модель уже не знаю, она была в расстройстве удалена. Да, eval_metrics всегда дефолтные использую.
Да, окей там получается меньше 20к деревьев, но больше 10к, разница небольшая. Но когда я грузил GPU модель(cell with load model), она долго грузилась, и оперативка подлетела в небеса

**Comment by annaveronika** [comment]
Непонятно сходу, что там за проблемы. Можешь обновиться, запустить обучение с logging_level='Info' и прислать вывод лога?
И проверь, полечились ли какие-то из проблем.

**Comment by JustM57** [comment]
@annaveronika если честно, сейчас немного не до гпу модели. Будет время - попробую
Кстати, вам нужен какой-нибудь кернел по работе со временными рядами? Я могу сделать по битку(конечно упрощённый с точки зрения фич)

**Comment by annaveronika** [comment]
Давай, было бы круто!

**Comment by Noxoomo** [comment]
с моделью воспроизвели проблему, будем фиксить

**Comment by JustM57** [comment]
@Noxoomo Здорово:) А графики к гпу никак не получается прикрутить?

**Comment by Noxoomo** [comment]
графики гпу проверим и поправим (не тензорбордные вообще должны работать, там видимо опечатка в именах файлов)

**Comment by Noxoomo** [comment]
да, тот, что качество у GPU странное было — ровно та жа проблема, что и размер

**Comment by JustM57** [comment]
@Noxoomo видимо перевод модели в момент окончания гпу подсчёта на цпу, был криво реализован? Т.е. посчиталась модель одна, но сохранилась совсем другая?

**Comment by Noxoomo** [comment]
при переезде на flatbuffer (пока гипотеза
, может и позже) была допущена ошибка в сохранении части модели.
Сейчас gpu-модели с кат-фичами сохраняются криво, если катфичей нету, то все работает

**Comment by Noxoomo** [comment]
Сохранение модели поправили, скоро приедет на гитхаб. На неделе выкатим пакет с исправлением

**Comment by Noxoomo** [comment]
А что с графиками? можно поподробнее, что конкретно не работает? (ipywidgets у меня графики отобразил при указании нужной директории)

**Comment by annaveronika** [comment]
closing because of inactivity. Please reopen or create new issue, if any problems are left.
